python/pyspark/rdd.py

```diff
@@ -2029,7 +2028,15 @@ def coalesce(self, numPartitions, shuffle=False):
         >>> sc.parallelize([1, 2, 3, 4, 5], 3).coalesce(1).glom().collect()
         [[1, 2, 3, 4, 5]]
         """
-        jrdd = self._jrdd.coalesce(numPartitions, shuffle)
+        if shuffle:
+            # In Scala's repartition code, we will distribute elements evenly across output
+            # partitions. However, the RDD from Python is serialized as a single binary data,
+            # so the distribution fails and produces highly skewed partitions. We need to
+            # convert it to a RDD of java object before repartitioning.
+            data_java_rdd = self._to_java_object_rdd().coalesce(numPartitions, shuffle)
```

Comment:
The reason that cause the skew should be large batch size, I think we could decrease the batch size to 10, then call repartition in JVM.

My worry is that _to_java_object_rdd() could be expensive, maybe we should have some benchmark for that.